{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "51075467-9f9c-4d1b-b84e-6b96911023f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Our imports\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72ee9398-dfda-4220-91bf-d64878d17f7e",
   "metadata": {},
   "source": [
    "# Clustering of TIM\n",
    "## What is the goal of this notebook? \n",
    "\n",
    "This notebook is intended to identify all unique chains in a dataset and their corresponding entity names. The results of this notebook can aid in the entity naming process by clustering all polymeric sequences in the dataset and retrieving the entity names associated with each member in each cluster.\n",
    "\n",
    "NOTE:\n",
    ">For clustering, we use **MMseqs2 coverage mode 1 (target mode)**.\n",
    ">ğŸ“– For more details, see the [MMseqs2 documentation](https://github.com/soedinglab/MMseqs2).\n",
    "\n",
    "Below is the workflow that this notebook follows. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74511f6c-02b0-4b6c-ae52-4b58bb512879",
   "metadata": {},
   "source": [
    "## Workflow\n",
    "\n",
    "This notebook follows a step-by-step process to cluster PDB entries based on their sequences and extract meaningful information from the results.\n",
    "\n",
    "1. __Setup__\n",
    "- The notebook begins by creating the necessary directories to store all files generated throughout the workflow.\n",
    "\n",
    "2. __Filtering SEQRES Entries__\n",
    "- Using the provided PDB IDs file, we match each ID to its corresponding entry in `seqres.txt`. We then generate a new FASTA file that contains only the SEQRES entries corresponding to the selected PDB IDs.\n",
    "\n",
    "3. __Clustering with MMseqs2__\n",
    "- The filtered FASTA file is clustered using __MMseqs2__. The output is a `.tsv` file containing information about the clusters, such as the representatives and members.\n",
    "\n",
    "4. __Generating Cluster FASTA Files__\n",
    "- Using the `.tsv` file, we generate individual FASTA files for each cluster. Each file contains the SEQRES entries for all members of a given cluster, including the representative sequence.\n",
    "\n",
    "5. __Extracting Entity Names__\n",
    "- For each cluster FASTA file, we extract the entity names. These names are compiled into a singular `.txt` file, along with the count of occurrences.\n",
    "    - Example entry of this file:\n",
    ">6BVE_B_cluster.txt{'TRIOSEPHOSPHATE ISOMERASE': 26}{'TRIOSEPHOSPHATE ISOMERASE': 26, 'PROTEIN (TRIOSEPHOSPHATE ISOMERASE)': 4}\n",
    "    - __6BVE_B_cluster.txt__ : The cluster representative\n",
    "    - __{'TRIOSEPHOSPHATE ISOMERASE': 26}__ : The most common present entity in the cluster\n",
    "    - __{'TRIOSEPHOSPHATE ISOMERASE': 26, 'PROTEIN (TRIOSEPHOSPHATE ISOMERASE)': 4}__ : All unique entities in the cluster"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f865274-7b7e-4138-85e2-c053edda9b6c",
   "metadata": {},
   "source": [
    "## Essential Files\n",
    "This notebook uses two essential files to run:\n",
    "1. `pdblist.txt`\n",
    "2. `seqres.txt`\n",
    "\n",
    "### PDBLIST File\n",
    "This is a txt file containing a list of  PDB IDs, with one ID per line.\n",
    "\n",
    "__Example:__\n",
    "- PDB_ID_1\n",
    "- PDB_ID_2\n",
    "- PDB_ID_3\n",
    "\n",
    "### SEQRES File\n",
    "`seqres.txt` is a FASTA-formatted file containing sequences for all entries in the PDB. This file is generated by running Step 1 in the tutorial notebooks for PDBCleanV2.\n",
    "\n",
    "__Example entry in seqres:__\n",
    ">102m_A mol:protein length:154  MYOGLOBIN\n",
    "MVLSEGEWQLVLHVWAKVEADVAGHGQDILIRLFKSHPETLEKFDRFKHLKTEAEMKASEDLKKAGVTVLTALGAILKKKGHHEAELKPLAQSHATKHKIPIKYLEFISEAIIHVLHSRHPGNFGADAQGAMNKALELFRKDIAAKYKELGYQG"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bfb731af-ca3b-4913-af89-8c6a4aacebcf",
   "metadata": {},
   "source": [
    "## Specify working directories\n",
    "1. Let's make a directory where we will store everything we make in this notebook. As well as specifying our working directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e243a92b-e666-45ab-9d13-0bf84881a1dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "PROJDIR = './TIM/'\n",
    "pdb_ids_path = PROJDIR + 'raw_bank/'\n",
    "output_directory = PROJDIR + 'clustering_storage/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3dd472fe-6261-4972-8d44-380ecc5f12f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Making the directory where we will store everything\n",
    "os.makedirs(output_directory, exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24d165b2-1e1a-44f5-81e7-308a871a0b53",
   "metadata": {},
   "source": [
    "* Lets save the file paths for both of our essential files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "424b2b1d-d3f8-4537-ac72-da7a03c02ed6",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdbid_file = PROJDIR + \"pdblist.txt\"\n",
    "seqres_file = PROJDIR + \"seqres.txt\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf13b10f-1d29-4f7f-b7e3-260db2317b04",
   "metadata": {},
   "source": [
    "2. We'll also create a directory within our primary directory to store files related to the coverage mode for MMSeqs2 that we'll work with.\n",
    "   - Mode 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "78cbe20c-9f71-46f9-8e09-5c8ea5db56c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "mode_directory = 'mode_1/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "58e73971-45f7-48b7-befb-15d946e7e07f",
   "metadata": {},
   "outputs": [],
   "source": [
    "mode_1_path = f'{output_directory}{mode_directory}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "896ee37d-a5e9-4e68-9272-c5272db1ae6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.makedirs(mode_1_path, exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed9f19ff-f646-40bc-b746-d67d38bde90c",
   "metadata": {},
   "source": [
    "3. We will then create a directory within our mode directory, which will store FASTA files we will generate from the outputted clusters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "81a29820-5d43-4450-bc69-18e42538bc5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_fasta_storage = 'cluster_files/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f953eda2-ec32-405b-baae-c17330abfac7",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_fasta_storage_path = f'{mode_1_path}/{cluster_fasta_storage}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "637acb3c-fb0b-41de-ad34-2e4ec71f5809",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.makedirs(cluster_fasta_storage_path, exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad474ec6-5f1e-40f6-9d88-31d241eaab2c",
   "metadata": {},
   "source": [
    "## How to create the 'pdblist.txt'\n",
    "\n",
    "We will run the following two commands on the `raw_bank` directory to obtain the PDB IDs we are interested in.\n",
    "\n",
    "1. `! ls path_to_directory | grep cif > pdbid_list_temp.txt`\n",
    "2. `! sed 's/\\.cif//g' pdbid_list_temp.txt > pdblist.txt`\n",
    "\n",
    "__Explanation:__\n",
    "- Line 1:\n",
    "    - This command lists all files in the specified directory whose names contain the string `'cif'`, and saves the list to 'pdbid_list_temp.txt'.\n",
    "\n",
    "- Line 2:\n",
    "    - This command __removes the `.cif` extension__ from each line in `pdbid_list_temp.txt` and writes the cleaned results (just the base PDB IDs) to `pdblist.txt`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e36b5d8a-b453-4a9f-8787-a47fbe87206b",
   "metadata": {},
   "outputs": [],
   "source": [
    "! ls {pdb_ids_path} | grep cif > pdbid_list_temp.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "2ea6819a-d30f-4269-83d1-3042f8209f18",
   "metadata": {},
   "outputs": [],
   "source": [
    "! sed 's/\\.cif//g' pdbid_list_temp.txt > pdblist.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "714fea40-bc1d-4902-9147-033ead044cc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "! rm pdbid_list_temp.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "720dfabc",
   "metadata": {},
   "outputs": [],
   "source": [
    "! mv pdblist.txt {PROJDIR}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e979808-3d3b-482d-bfb3-10a3ca8eada3",
   "metadata": {},
   "source": [
    "## Editing SEQRES\n",
    "The entity names in the SEQRES file contain inconsistencies, including the use of special characters such as single quotation marks and asterisks. These irregularities can impact our clustering results. To address this, we will standardize the SEQRES entries by removing such inconsistencies."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "4c0f2137-795f-4af6-a4f1-5de70e1deecc",
   "metadata": {},
   "outputs": [],
   "source": [
    "output_seqres_path = PROJDIR + 'seqres2.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "964a7604-563e-49b7-936b-04c8259fdb14",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(seqres_file, 'r') as input_file, open(output_seqres_path, 'w') as output_file:\n",
    "    for meta_line, sequence_line in zip(input_file, input_file):\n",
    "        \n",
    "        # Clean up the line and split into parts\n",
    "        attributes = meta_line.strip().split()\n",
    "        \n",
    "        # Obtain the PDB ID and Chain ID, then only capitalize the PDB ID\n",
    "        pdbid, chid = attributes[0].split('_')\n",
    "        pdb_chid = f'{pdbid.upper()}_{chid}'\n",
    "        \n",
    "        # Capitalize all other attributes, and replace special characters from entity name\n",
    "        set_attributes = ' '.join(attr.upper() for attr in attributes[1:3])\n",
    "        updated_entity_name =  ' '.join(word.upper() for word in attributes[3:]).replace(\"'\", \"\").replace(\"*\",\"-\")\n",
    "        \n",
    "        #re-construct new meta-line with our updated attributes\n",
    "        new_meta_line = f'{pdb_chid} {set_attributes} {updated_entity_name}\\n'\n",
    "        # Clean up the line and split into parts\n",
    "        \n",
    "        # Write to output\n",
    "        output_file.write(new_meta_line)\n",
    "        output_file.write(sequence_line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "aec1c77a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Replaces our old seqres file\n",
    "! mv {PROJDIR}/seqres2.txt {PROJDIR}/seqres.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9e9961d-a9dc-4dd7-8ffb-9abcf9fe2a35",
   "metadata": {},
   "source": [
    "## PDB IDs File to FASTA\n",
    "We start with our **PDB IDs** file, and match the PDB IDs in the file to their corresponding entries in the `seqres` file.  \n",
    "\n",
    "### Define functions\n",
    "Weâ€™ll define a few __helper functions__ essential for performing this task:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "ef7aac25-16a7-4608-9c35-985310f4bd29",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_attributes_seqres(file_path, attribute='PDBID_CHID'):\n",
    "    \"\"\"\n",
    "    Extracts the specified attributes from the metadata line of a FASTA file.\n",
    "\n",
    "    file_path(str) : File path leading to seqres.txt\n",
    "    attribute(str) : Metadata attribute to extract. Options are 'ENTITY_LABEL',\n",
    "                    'PDBID', 'CHID', 'PDBID_CHID'. The default is PDBID_CHID.\n",
    "    Returns : \n",
    "        attribute_list(list) : List of extracted attributes.\n",
    "    \"\"\"\n",
    "    attribute_list = []\n",
    "    attribute = attribute.upper()\n",
    "\n",
    "    # Had to add the encoding parameter due to an error that kept occuring when reading the file.\n",
    "    with open(file_path, 'r', encoding='latin-1') as seqres:\n",
    "        for i, line in enumerate(seqres):\n",
    "            \n",
    "            if i % 2 == 0:\n",
    "                \n",
    "                if attribute == 'ENTITY_LABEL':\n",
    "                    entity_name = ' '.join(line.split()[3:])\n",
    "                    attribute_list.append(entity_name.upper())\n",
    "                # Every other attribute outside of entity label is found within the first item \n",
    "                # obtained after we perform the split function on the metadataline,\n",
    "                # Therefore we perform the following code after the 'else' function.\n",
    "                \n",
    "                else:\n",
    "                    line = line.split()[0][1:]\n",
    "                    if attribute == 'PDBID':\n",
    "                        # Everything before the '_' is the PDB ID]\n",
    "                        attribute_list.append(line.split('_')[0].upper())\n",
    "                # Everything after the '_' is the chain ID\n",
    "                    elif attribute == 'CHID':\n",
    "                        attribute_list.append(line.split('_')[1])\n",
    "                    elif attribute == 'PDBID_CHID':\n",
    "                        attribute_list.append(line)\n",
    "            \n",
    "            elif i % 2 == 1 and attribute == 'SEQUENCE':\n",
    "                attribute_list.append(line.strip().upper())\n",
    "                    \n",
    "    return attribute_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "0e3f45bb-575c-42b2-a9f1-ad3e9090154b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def seqres_entry_dict(file_path):\n",
    "    \"\"\"\n",
    "    Returns a dictionary with the keys being the PDB entry ID and chain ID (e.g. pdbID_chID) and the values\n",
    "    as [meta data, sequence].\n",
    "\n",
    "    file_path : file path leading to seqres\n",
    "\n",
    "    Note : The chain ID is case sensitive, below we make sure that the \n",
    "    \"\"\"\n",
    "    entry_dict = {}\n",
    "    with open(file_path, 'r') as file:\n",
    "        for meta_line, sequence_line in zip(file, file):\n",
    "            \n",
    "            # Here we capitalize everything in pdb_chid except for the chain id\n",
    "            pdb_chid = meta_line.split()[0][1:]\n",
    "            key = pdb_chid\n",
    "\n",
    "            # We obtain all other components in our metaline (e.g mol, length, entity name)\n",
    "            rest_line_cap = [word.upper() for word in meta_line.split()[1:]]\n",
    "            \n",
    "            # We construct our metaline with our capitalized components\n",
    "            meta_line = '>' + pdb_chid + ' ' + ' '.join(rest_line_cap) + '\\n'\n",
    "\n",
    "            # We make our dictionary from our capitalzed seqres components\n",
    "            entry_dict[key] = [meta_line, sequence_line.upper()]\n",
    "\n",
    "    return entry_dict "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d89d07af-316b-44f7-b1d4-779793231e7a",
   "metadata": {},
   "source": [
    "- **Task**: Here we save the `PDB ID`'s from the provided txt file containing only PDB IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "20ad439c-3d79-44cb-9744-a734a8b0ad2c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "An attribute from our list: 1AG1\n"
     ]
    }
   ],
   "source": [
    "PDB_ID_list = []\n",
    "with open (pdbid_file, 'r') as file:\n",
    "    for line in file:\n",
    "        PDB_ID_list.append(line[:4].upper())\n",
    "\n",
    "test_pdbid = PDB_ID_list[0]\n",
    "print(f'An attribute from our list: {PDB_ID_list[PDB_ID_list.index(test_pdbid)]}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed2410b8-c697-4726-8b27-77aea3c57b3f",
   "metadata": {},
   "source": [
    "- **Task**: Here we extract the `pdbID_chID` labels from `seqres` and store them in a __list__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "aa4f6066-0d58-417c-a4b8-b172cd07df69",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "An attribute from our list: 1AG1_O\n"
     ]
    }
   ],
   "source": [
    "pdb_chid_list = get_attributes_seqres(seqres_file, 'PDBID_CHID')\n",
    "\n",
    "test_pdbid2 = [pdb for pdb in pdb_chid_list if pdb[:4] == test_pdbid][0]\n",
    "print(f'An attribute from our list: {pdb_chid_list[pdb_chid_list.index(test_pdbid2)]}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66cecbb5-4793-4172-8b4b-a2a95bfc5d96",
   "metadata": {},
   "source": [
    "Below we extract the matching `pdbID_chID` from `seqres` that corresponds to the `PDB ID` in the `pdb_id` file, and store it in a list.  \n",
    "\n",
    "- **Example**:  \n",
    "  - `PDB ID`: `1AG1`  \n",
    "  - `SEQRES pdbID_chID`: `1AG1_A`  \n",
    "\n",
    "Since the PDB IDs match, we store the `SEQRES pdbID_chID` (e.g., `1AG1_A`) in a list.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "e2f907ba-fee5-488f-b6ec-43ff11718059",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "An attribute from our list: 1AG1_O\n"
     ]
    }
   ],
   "source": [
    "matching_labels = [label for label in pdb_chid_list if label.split('_', 1)[0] in PDB_ID_list]\n",
    "print(f'An attribute from our list: {matching_labels[matching_labels.index(test_pdbid2)]}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71fd8255-a889-48b1-b767-ea1cb5affe2d",
   "metadata": {},
   "source": [
    "- **Task**: Below we make a dictionary out of our `seqres` file. (Read function documentation for further clarity)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "8ea4bc30-c54d-4579-a327-d7691d23bed0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "An attribute from our dictionary: ['>1AG1_O MOL:PROTEIN LENGTH:250 TRIOSEPHOSPHATE ISOMERASE\\n', 'MSKPQPIAAANWKCNGSQQSLSELIDLFNSTSINHDVQCVVASTFVHLAMTKERLSHPKFVIAAQNAIAKSGAFTGEVSLPILKDFGVNWIVLGHSERRAYYGETNEIVADKVAAAVASGFMVIACIGETLQERESGRTAVVVLTQIAAIAKKLKKADWAKVVIAYEPVWAIGTGKVATPQQAQEAHALIRSWVSSKIGADVAGELRILYGGSVNGKNARTLYQQRDVNGFLVGGASLKPEFVDIIKATQ\\n']\n"
     ]
    }
   ],
   "source": [
    "seqres_dict = seqres_entry_dict(seqres_file)\n",
    "print(f'An attribute from our dictionary: {seqres_dict[test_pdbid2]}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9e98dd7-823a-4a1f-86c1-3f2f63c29e4f",
   "metadata": {},
   "source": [
    "- **Task**: with all of our components, we now can write a FASTA file containing only the structures we are interested in.\n",
    "- **Output**: A FASTA file within our mode_1 directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "0a23fa00-b15d-4690-8c79-1142fae61f75",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specifying what I wish for my FASTA file to be called\n",
    "fasta_output_name = 'tim_seqres.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "b87d633a-6006-4549-928a-928b69e473ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(f'{mode_1_path}/{fasta_output_name}', 'w') as file:\n",
    "    for label in matching_labels:\n",
    "        lines_to_write = seqres_dict[label]\n",
    "        if lines_to_write:\n",
    "            file.writelines(lines_to_write)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "1d4b6593-9554-4c57-a830-560d9dff48c2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The first 5 entries of our 'tim_seqres.txt'\n",
      "\n",
      ">1AG1_O MOL:PROTEIN LENGTH:250 TRIOSEPHOSPHATE ISOMERASE\r\n",
      "MSKPQPIAAANWKCNGSQQSLSELIDLFNSTSINHDVQCVVASTFVHLAMTKERLSHPKFVIAAQNAIAKSGAFTGEVSLPILKDFGVNWIVLGHSERRAYYGETNEIVADKVAAAVASGFMVIACIGETLQERESGRTAVVVLTQIAAIAKKLKKADWAKVVIAYEPVWAIGTGKVATPQQAQEAHALIRSWVSSKIGADVAGELRILYGGSVNGKNARTLYQQRDVNGFLVGGASLKPEFVDIIKATQ\r\n",
      ">1AG1_T MOL:PROTEIN LENGTH:250 TRIOSEPHOSPHATE ISOMERASE\r\n",
      "MSKPQPIAAANWKCNGSQQSLSELIDLFNSTSINHDVQCVVASTFVHLAMTKERLSHPKFVIAAQNAIAKSGAFTGEVSLPILKDFGVNWIVLGHSERRAYYGETNEIVADKVAAAVASGFMVIACIGETLQERESGRTAVVVLTQIAAIAKKLKKADWAKVVIAYEPVWAIGTGKVATPQQAQEAHALIRSWVSSKIGADVAGELRILYGGSVNGKNARTLYQQRDVNGFLVGGASLKPEFVDIIKATQ\r\n",
      ">1AW1_A MOL:PROTEIN LENGTH:256 TRIOSEPHOSPHATE ISOMERASE\r\n",
      "MRHPVVMGNWKLNGSKEMVVDLLNGLNAELEGVTGVDVAVAPPALFVDLAERTLTEAGSAIILGAQNTDLNNSGAFTGDMSPAMLKEFGATHIIIGHSERREYHAESDEFVAKKFAFLKENGLTPVLCIGESDAQNEAGETMAVCARQLDAVINTQGVEALEGAIIAYEPIWAIGTGKAATAEDAQRIHAQIRAHIAEKSEAVAKNVVIQYGGSVKPENAAAYFAQPDIDGALVGGAALDAKSFAAIAKAAAEAKA\r\n",
      ">1AW1_B MOL:PROTEIN LENGTH:256 TRIOSEPHOSPHATE ISOMERASE\r\n",
      "MRHPVVMGNWKLNGSKEMVVDLLNGLNAELEGVTGVDVAVAPPALFVDLAERTLTEAGSAIILGAQNTDLNNSGAFTGDMSPAMLKEFGATHIIIGHSERREYHAESDEFVAKKFAFLKENGLTPVLCIGESDAQNEAGETMAVCARQLDAVINTQGVEALEGAIIAYEPIWAIGTGKAATAEDAQRIHAQIRAHIAEKSEAVAKNVVIQYGGSVKPENAAAYFAQPDIDGALVGGAALDAKSFAAIAKAAAEAKA\r\n",
      ">1AW1_D MOL:PROTEIN LENGTH:256 TRIOSEPHOSPHATE ISOMERASE\r\n",
      "MRHPVVMGNWKLNGSKEMVVDLLNGLNAELEGVTGVDVAVAPPALFVDLAERTLTEAGSAIILGAQNTDLNNSGAFTGDMSPAMLKEFGATHIIIGHSERREYHAESDEFVAKKFAFLKENGLTPVLCIGESDAQNEAGETMAVCARQLDAVINTQGVEALEGAIIAYEPIWAIGTGKAATAEDAQRIHAQIRAHIAEKSEAVAKNVVIQYGGSVKPENAAAYFAQPDIDGALVGGAALDAKSFAAIAKAAAEAKA\r\n"
     ]
    }
   ],
   "source": [
    "print(\"The first 5 entries of our 'tim_seqres.txt'\\n\")\n",
    "! head {mode_1_path}/{fasta_output_name}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b15873de-ce25-4c83-b451-e54fbf8acbfe",
   "metadata": {},
   "source": [
    "# Performing Clustering\n",
    "Using the program MMSeqs2, we will perform the clustering of our created FASTA file.\n",
    "\n",
    "## MMseqs2 clustering steps\n",
    "I will run the following three lines:\n",
    "1. __mmseqs createdb fasta_file.txt pdbDB__\n",
    "   - Converts the input sequence file seqres.txt into an MMseqs2 database named pdbDB (pdbDB was a name of my choosing)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "79e5093c-89cd-4a39-829b-70e29460e2ec",
   "metadata": {},
   "source": [
    "2. __mmseqs cluster pdbDB clusterDB tmp --min-seq-id 0.95 --cluster-mode 2 --cov-mode 1 -c 0.8__\n",
    "   - Clusters the sequences in `pdbDB` based on a **minimum sequence identity** (`--min-seq-id`) and **coverage** (`-c`).  \n",
    "   - The example values here are `0.95` for sequence identity and `0.8` for coverage, but you can **choose any values** suitable for your analysis. We also use `clustering mode 2` and `coverage mode 1`\n",
    "   - Outputs the clustering result into `clusterDB`."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf71cdfc-5306-4c2b-8cad-d7bdac11d564",
   "metadata": {},
   "source": [
    "3. __mmseqs createtsv pdbDB pdbDB clusterDB clusters.tsv__\n",
    "   - Outputs the clusters from clusterDB into a readable TSV file (clusters.tsv)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "f8e3a18f-1761-46dd-859d-5765d44dff0e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# For a good part of this notebook, we will be working within this directory\n",
    "os.chdir(mode_1_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "3dd802c0-92b9-429c-b2b3-8528fee808aa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "createdb tim_seqres.txt pdbDB \n",
      "\n",
      "MMseqs Version:       \t14.7e284\n",
      "Database type         \t0\n",
      "Shuffle input database\ttrue\n",
      "Createdb mode         \t0\n",
      "Write lookup file     \t1\n",
      "Offset of numeric ids \t0\n",
      "Compressed            \t0\n",
      "Verbosity             \t3\n",
      "\n",
      "Converting sequences\n",
      "[607] 0s 6ms\n",
      "Time for merging to pdbDB_h: 0h 0m 0s 3ms\n",
      "Time for merging to pdbDB: 0h 0m 0s 2ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 23ms\n"
     ]
    }
   ],
   "source": [
    "!mmseqs createdb {fasta_output_name} pdbDB "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "5dea401a-e852-4448-b680-890b0a7f3a4b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Create directory tmp\n",
      "cluster pdbDB clusterDB tmp --min-seq-id 0.95 --cluster-mode 2 --cov-mode 1 -c 0.8 \n",
      "\n",
      "MMseqs Version:                     \t14.7e284\n",
      "Substitution matrix                 \taa:blosum62.out,nucl:nucleotide.out\n",
      "Seed substitution matrix            \taa:VTML80.out,nucl:nucleotide.out\n",
      "Sensitivity                         \t4\n",
      "k-mer length                        \t0\n",
      "k-score                             \tseq:2147483647,prof:2147483647\n",
      "Alphabet size                       \taa:21,nucl:5\n",
      "Max sequence length                 \t65535\n",
      "Max results per query               \t20\n",
      "Split database                      \t0\n",
      "Split mode                          \t2\n",
      "Split memory limit                  \t0\n",
      "Coverage threshold                  \t0.8\n",
      "Coverage mode                       \t1\n",
      "Compositional bias                  \t1\n",
      "Compositional bias                  \t1\n",
      "Diagonal scoring                    \ttrue\n",
      "Exact k-mer matching                \t0\n",
      "Mask residues                       \t1\n",
      "Mask residues probability           \t0.9\n",
      "Mask lower case residues            \t0\n",
      "Minimum diagonal score              \t15\n",
      "Selected taxa                       \t\n",
      "Include identical seq. id.          \tfalse\n",
      "Spaced k-mers                       \t1\n",
      "Preload mode                        \t0\n",
      "Pseudo count a                      \tsubstitution:1.100,context:1.400\n",
      "Pseudo count b                      \tsubstitution:4.100,context:5.800\n",
      "Spaced k-mer pattern                \t\n",
      "Local temporary path                \t\n",
      "Threads                             \t8\n",
      "Compressed                          \t0\n",
      "Verbosity                           \t3\n",
      "Add backtrace                       \tfalse\n",
      "Alignment mode                      \t3\n",
      "Alignment mode                      \t0\n",
      "Allow wrapped scoring               \tfalse\n",
      "E-value threshold                   \t0.001\n",
      "Seq. id. threshold                  \t0.95\n",
      "Min alignment length                \t0\n",
      "Seq. id. mode                       \t0\n",
      "Alternative alignments              \t0\n",
      "Max reject                          \t2147483647\n",
      "Max accept                          \t2147483647\n",
      "Score bias                          \t0\n",
      "Realign hits                        \tfalse\n",
      "Realign score bias                  \t-0.2\n",
      "Realign max seqs                    \t2147483647\n",
      "Correlation score weight            \t0\n",
      "Gap open cost                       \taa:11,nucl:5\n",
      "Gap extension cost                  \taa:1,nucl:2\n",
      "Zdrop                               \t40\n",
      "Rescore mode                        \t0\n",
      "Remove hits by seq. id. and coverage\tfalse\n",
      "Sort results                        \t0\n",
      "Cluster mode                        \t2\n",
      "Max connected component depth       \t1000\n",
      "Similarity type                     \t2\n",
      "Single step clustering              \tfalse\n",
      "Cascaded clustering steps           \t3\n",
      "Cluster reassign                    \tfalse\n",
      "Remove temporary files              \tfalse\n",
      "Force restart with latest tmp       \tfalse\n",
      "MPI runner                          \t\n",
      "k-mers per sequence                 \t21\n",
      "Scale k-mers per sequence           \taa:0.000,nucl:0.200\n",
      "Adjust k-mer length                 \tfalse\n",
      "Shift hash                          \t67\n",
      "Include only extendable             \tfalse\n",
      "Skip repeating k-mers               \tfalse\n",
      "\n",
      "Set cluster sensitivity to -s 1.000000\n",
      "Set cluster iterations to 1\n",
      "linclust pdbDB tmp/245840764643801654/clu_redundancy tmp/245840764643801654/linclust --cluster-mode 2 --max-iterations 1000 --similarity-type 2 --threads 8 --compressed 0 -v 3 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.95 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 0 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --alph-size aa:13,nucl:5 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 -k 0 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --rescore-mode 0 --filter-hits 0 --sort-results 0 --remove-tmp-files 0 --force-reuse 0 \n",
      "\n",
      "kmermatcher pdbDB tmp/245840764643801654/linclust/11734362224902840304/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.95 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 1 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "kmermatcher pdbDB tmp/245840764643801654/linclust/11734362224902840304/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.95 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 1 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Database size: 675 type: Aminoacid\n",
      "Reduced amino acid alphabet: (A S T) (C) (D B N) (E Q Z) (F Y) (G) (H) (I V) (K R) (L J M) (P) (W) (X) \n",
      "\n",
      "Generate k-mers list for 1 split\n",
      "[=================================================================] 100.00% 675 0s 4ms      \n",
      "Sort kmer 0h 0m 0s 0ms\n",
      "Sort by rep. sequence 0h 0m 0s 0ms\n",
      "Time for fill: 0h 0m 0s 0ms\n",
      "Time for merging to pref: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 16ms\n",
      "rescorediagonal pdbDB pdbDB tmp/245840764643801654/linclust/11734362224902840304/pref tmp/245840764643801654/linclust/11734362224902840304/pref_rescore1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 0 --wrapped-scoring 0 --filter-hits 0 -e 0.001 -c 0.8 -a 0 --cov-mode 1 --min-seq-id 0.95 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 675 0s 3ms      \n",
      "Time for merging to pref_rescore1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 11ms\n",
      "clust pdbDB tmp/245840764643801654/linclust/11734362224902840304/pref_rescore1 tmp/245840764643801654/linclust/11734362224902840304/pre_clust --cluster-mode 2 --max-iterations 1000 --similarity-type 2 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Clustering mode: Greedy\n",
      "Total time: 0h 0m 0s 13ms\n",
      "\n",
      "Size of the sequence database: 675\n",
      "Size of the alignment database: 675\n",
      "Number of clusters: 85\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to pre_clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 18ms\n",
      "createsubdb tmp/245840764643801654/linclust/11734362224902840304/order_redundancy pdbDB tmp/245840764643801654/linclust/11734362224902840304/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "createsubdb tmp/245840764643801654/linclust/11734362224902840304/order_redundancy tmp/245840764643801654/linclust/11734362224902840304/pref tmp/245840764643801654/linclust/11734362224902840304/pref_filter1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to pref_filter1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "filterdb tmp/245840764643801654/linclust/11734362224902840304/pref_filter1 tmp/245840764643801654/linclust/11734362224902840304/pref_filter2 --filter-file tmp/245840764643801654/linclust/11734362224902840304/order_redundancy --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Filtering using file(s)\n",
      "[=================================================================] 100.00% 85 0s 1ms      \n",
      "Time for merging to pref_filter2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 7ms\n",
      "rescorediagonal tmp/245840764643801654/linclust/11734362224902840304/input_step_redundancy tmp/245840764643801654/linclust/11734362224902840304/input_step_redundancy tmp/245840764643801654/linclust/11734362224902840304/pref_filter2 tmp/245840764643801654/linclust/11734362224902840304/pref_rescore2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 1 --wrapped-scoring 0 --filter-hits 1 -e 0.001 -c 0.8 -a 0 --cov-mode 1 --min-seq-id 0.95 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 85 0s 1ms      \n",
      "Time for merging to pref_rescore2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 10ms\n",
      "align tmp/245840764643801654/linclust/11734362224902840304/input_step_redundancy tmp/245840764643801654/linclust/11734362224902840304/input_step_redundancy tmp/245840764643801654/linclust/11734362224902840304/pref_rescore2 tmp/245840764643801654/linclust/11734362224902840304/aln --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.95 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 0 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 8 --compressed 0 -v 3 \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Compute score, coverage and sequence identity\r\n",
      "Query database size: 85 type: Aminoacid\r\n",
      "Target database size: 85 type: Aminoacid\r\n",
      "Calculation of alignments\r\n",
      "[===>                                                             ] 4.76% 5 eta 0s       \r",
      "[>                                                                ] 0.00% 1 eta -\r",
      "[=>                                                               ] 2.38% 3 eta 0s       \r",
      "[===>                                                             ] 5.95% 6 eta 0s       \r",
      "[==>                                                              ] 3.57% 4 eta 0s       \r",
      "[>                                                                ] 1.19% 2 eta 0s       \r",
      "[=====>                                                           ] 8.33% 8 eta 0s       \r",
      "[======>                                                          ] 9.52% 9 eta 0s       \r",
      "[=======>                                                         ] 11.90% 11 eta 0s       \r",
      "[======>                                                          ] 10.71% 10 eta 0s       \r",
      "[========>                                                        ] 13.10% 12 eta 0s       \r",
      "[====>                                                            ] 7.14% 7 eta 0s       \r",
      "[=========>                                                       ] 14.29% 13 eta 0s       \r",
      "[==========>                                                      ] 15.48% 14 eta 0s       \r",
      "[==========>                                                      ] 16.67% 15 eta 0s       \r",
      "[===========>                                                     ] 17.86% 16 eta 0s       \r",
      "[============>                                                    ] 19.05% 17 eta 0s       \r",
      "[=============>                                                   ] 21.43% 19 eta 0s       \r",
      "[=============>                                                   ] 20.24% 18 eta 0s       \r",
      "[==============>                                                  ] 22.62% 20 eta 0s       \r",
      "[===============>                                                 ] 23.81% 21 eta 0s       \r",
      "[=================>                                               ] 27.38% 24 eta 0s       \r",
      "[================>                                                ] 25.00% 22 eta 0s       \r",
      "[=================>                                               ] 26.19% 23 eta 0s       \r",
      "[==================>                                              ] 28.57% 25 eta 0s       \r",
      "[====================>                                            ] 30.95% 27 eta 0s       \r",
      "[===================>                                             ] 29.76% 26 eta 0s       \r",
      "[=====================>                                           ] 33.33% 29 eta 0s       \r",
      "[======================>                                          ] 34.52% 30 eta 0s       \r",
      "[====================>                                            ] 32.14% 28 eta 0s       \r",
      "[=======================>                                         ] 35.71% 31 eta 0s       \r",
      "[=======================>                                         ] 36.90% 32 eta 0s       \r",
      "[========================>                                        ] 38.10% 33 eta 0s       \r",
      "[==========================>                                      ] 40.48% 35 eta 0s       \r",
      "[===========================>                                     ] 42.86% 37 eta 0s       \r",
      "[===========================>                                     ] 41.67% 36 eta 0s       \r",
      "[============================>                                    ] 44.05% 38 eta 0s       \r",
      "[=========================>                                       ] 39.29% 34 eta 0s       \r",
      "[==============================>                                  ] 46.43% 40 eta 0s       \r",
      "[==============================>                                  ] 47.62% 41 eta 0s       \r",
      "[=============================>                                   ] 45.24% 39 eta 0s       \r",
      "[===============================>                                 ] 48.81% 42 eta 0s       \r",
      "[================================>                                ] 50.00% 43 eta 0s       \r",
      "[=================================>                               ] 51.19% 44 eta 0s       \r",
      "[====================================>                            ] 55.95% 48 eta 0s       \r",
      "[==================================>                              ] 53.57% 46 eta 0s       \r",
      "[==================================>                              ] 52.38% 45 eta 0s       \r",
      "[=====================================>                           ] 58.33% 50 eta 0s       \r",
      "[===================================>                             ] 54.76% 47 eta 0s       \r",
      "[=====================================>                           ] 57.14% 49 eta 0s       \r",
      "[======================================>                          ] 59.52% 51 eta 0s       \r",
      "[=======================================>                         ] 60.71% 52 eta 0s       \r",
      "[=========================================>                       ] 64.29% 55 eta 0s       \r",
      "[========================================>                        ] 61.90% 53 eta 0s       \r",
      "[=========================================>                       ] 63.10% 54 eta 0s       \r",
      "[==========================================>                      ] 65.48% 56 eta 0s       \r",
      "[============================================>                    ] 67.86% 58 eta 0s       \r",
      "[===============================================>                 ] 72.62% 62 eta 0s       \r",
      "[===========================================>                     ] 66.67% 57 eta 0s       \r",
      "[==============================================>                  ] 71.43% 61 eta 0s       \r",
      "[=============================================>                   ] 70.24% 60 eta 0s       \r",
      "[=================================================>               ] 76.19% 65 eta 0s       \r",
      "[============================================>                    ] 69.05% 59 eta 0s       \r",
      "[===============================================>                 ] 73.81% 63 eta 0s       \r",
      "[================================================>                ] 75.00% 64 eta 0s       \r",
      "[==================================================>              ] 77.38% 66 eta 0s       \r",
      "[====================================================>            ] 80.95% 69 eta 0s       \r",
      "[=====================================================>           ] 82.14% 70 eta 0s       \r",
      "[===================================================>             ] 78.57% 67 eta 0s       \r",
      "[========================================================>        ] 86.90% 74 eta 0s       \r",
      "[======================================================>          ] 83.33% 71 eta 0s       \r",
      "[=========================================================>       ] 88.10% 75 eta 0s       \r",
      "[===================================================>             ] 79.76% 68 eta 0s       \r",
      "[======================================================>          ] 84.52% 72 eta 0s       \r",
      "[==========================================================>      ] 90.48% 77 eta 0s       \r",
      "[==========================================================>      ] 89.29% 76 eta 0s       \r",
      "[===========================================================>     ] 91.67% 78 eta 0s       \r",
      "[=======================================================>         ] 85.71% 73 eta 0s       \r",
      "[============================================================>    ] 92.86% 79 eta 0s       \r",
      "[=============================================================>   ] 94.05% 80 eta 0s       \r",
      "[==============================================================>  ] 96.43% 82 eta 0s       \r",
      "[=============================================================>   ] 95.24% 81 eta 0s       \r",
      "[===============================================================> ] 97.62% 83 eta 0s       \r",
      "[================================================================>] 98.81% 84 eta 0s       \r",
      "[=================================================================] 100.00% 85 0s 2ms\r\n",
      "Time for merging to aln: 0h 0m 0s 0ms\r\n",
      "94 alignments calculated\r\n",
      "87 sequence pairs passed the thresholds (0.925532 of overall calculated)\r\n",
      "1.023529 hits per query sequence\r\n",
      "Time for processing: 0h 0m 0s 9ms\r\n",
      "clust tmp/245840764643801654/linclust/11734362224902840304/input_step_redundancy tmp/245840764643801654/linclust/11734362224902840304/aln tmp/245840764643801654/linclust/11734362224902840304/clust --cluster-mode 2 --max-iterations 1000 --similarity-type 2 --threads 8 --compressed 0 -v 3 \r\n",
      "\r\n",
      "Clustering mode: Greedy\r\n",
      "Total time: 0h 0m 0s 0ms\r\n",
      "\r\n",
      "Size of the sequence database: 85\r\n",
      "Size of the alignment database: 85\r\n",
      "Number of clusters: 83\r\n",
      "\r\n",
      "Writing results 0h 0m 0s 0ms\r\n",
      "Time for merging to clust: 0h 0m 0s 0ms\r\n",
      "Time for processing: 0h 0m 0s 4ms\r\n",
      "mergeclusters pdbDB tmp/245840764643801654/clu_redundancy tmp/245840764643801654/linclust/11734362224902840304/pre_clust tmp/245840764643801654/linclust/11734362224902840304/clust --threads 8 --compressed 0 -v 3 \r\n",
      "\r\n",
      "Clustering step 1\r\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[>                                                                ] 0.00% 1 eta -\r",
      "[>                                                                ] 1.19% 2 eta 0s       \r",
      "[=>                                                               ] 2.38% 3 eta 0s       \r",
      "[==>                                                              ] 3.57% 4 eta 0s       \r",
      "[===>                                                             ] 4.76% 5 eta 0s       \r",
      "[===>                                                             ] 5.95% 6 eta 0s       \r",
      "[====>                                                            ] 7.14% 7 eta 0s       \r",
      "[=====>                                                           ] 8.33% 8 eta 0s       \r",
      "[======>                                                          ] 9.52% 9 eta 0s       \r",
      "[======>                                                          ] 10.71% 10 eta 0s       \r",
      "[=======>                                                         ] 11.90% 11 eta 0s       \r",
      "[========>                                                        ] 13.10% 12 eta 0s       \r",
      "[=========>                                                       ] 14.29% 13 eta 0s       \r",
      "[==========>                                                      ] 15.48% 14 eta 0s       \r",
      "[==========>                                                      ] 16.67% 15 eta 0s       \r",
      "[===========>                                                     ] 17.86% 16 eta 0s       \r",
      "[============>                                                    ] 19.05% 17 eta 0s       \r",
      "[=============>                                                   ] 20.24% 18 eta 0s       \r",
      "[=============>                                                   ] 21.43% 19 eta 0s       \r",
      "[==============>                                                  ] 22.62% 20 eta 0s       \r",
      "[===============>                                                 ] 23.81% 21 eta 0s       \r",
      "[================>                                                ] 25.00% 22 eta 0s       \r",
      "[=================>                                               ] 26.19% 23 eta 0s       \r",
      "[=================>                                               ] 27.38% 24 eta 0s       \r",
      "[==================>                                              ] 28.57% 25 eta 0s       \r",
      "[===================>                                             ] 29.76% 26 eta 0s       \r",
      "[====================>                                            ] 30.95% 27 eta 0s       \r",
      "[====================>                                            ] 32.14% 28 eta 0s       \r",
      "[=====================>                                           ] 33.33% 29 eta 0s       \r",
      "[======================>                                          ] 34.52% 30 eta 0s       \r",
      "[=======================>                                         ] 35.71% 31 eta 0s       \r",
      "[=======================>                                         ] 36.90% 32 eta 0s       \r",
      "[========================>                                        ] 38.10% 33 eta 0s       \r",
      "[=========================>                                       ] 39.29% 34 eta 0s       \r",
      "[==========================>                                      ] 40.48% 35 eta 0s       \r",
      "[===========================>                                     ] 41.67% 36 eta 0s       \r",
      "[===========================>                                     ] 42.86% 37 eta 0s       \r",
      "[============================>                                    ] 44.05% 38 eta 0s       \r",
      "[=============================>                                   ] 45.24% 39 eta 0s       \r",
      "[==============================>                                  ] 46.43% 40 eta 0s       \r",
      "[==============================>                                  ] 47.62% 41 eta 0s       \r",
      "[===============================>                                 ] 48.81% 42 eta 0s       \r",
      "[================================>                                ] 50.00% 43 eta 0s       \r",
      "[=================================>                               ] 51.19% 44 eta 0s       \r",
      "[==================================>                              ] 52.38% 45 eta 0s       \r",
      "[==================================>                              ] 53.57% 46 eta 0s       \r",
      "[===================================>                             ] 54.76% 47 eta 0s       \r",
      "[====================================>                            ] 55.95% 48 eta 0s       \r",
      "[=====================================>                           ] 57.14% 49 eta 0s       \r",
      "[=====================================>                           ] 58.33% 50 eta 0s       \r",
      "[======================================>                          ] 59.52% 51 eta 0s       \r",
      "[=======================================>                         ] 60.71% 52 eta 0s       \r",
      "[========================================>                        ] 61.90% 53 eta 0s       \r",
      "[=========================================>                       ] 63.10% 54 eta 0s       \r",
      "[=========================================>                       ] 64.29% 55 eta 0s       \r",
      "[==========================================>                      ] 65.48% 56 eta 0s       \r",
      "[===========================================>                     ] 66.67% 57 eta 0s       \r",
      "[============================================>                    ] 67.86% 58 eta 0s       \r",
      "[============================================>                    ] 69.05% 59 eta 0s       \r",
      "[=============================================>                   ] 70.24% 60 eta 0s       \r",
      "[==============================================>                  ] 71.43% 61 eta 0s       \r",
      "[===============================================>                 ] 72.62% 62 eta 0s       \r",
      "[===============================================>                 ] 73.81% 63 eta 0s       \r",
      "[================================================>                ] 75.00% 64 eta 0s       \r",
      "[=================================================>               ] 76.19% 65 eta 0s       \r",
      "[==================================================>              ] 77.38% 66 eta 0s       \r",
      "[===================================================>             ] 78.57% 67 eta 0s       \r",
      "[===================================================>             ] 79.76% 68 eta 0s       \r",
      "[====================================================>            ] 80.95% 69 eta 0s       \r",
      "[=====================================================>           ] 82.14% 70 eta 0s       \r",
      "[======================================================>          ] 83.33% 71 eta 0s       \r",
      "[======================================================>          ] 84.52% 72 eta 0s       \r",
      "[=======================================================>         ] 85.71% 73 eta 0s       \r",
      "[========================================================>        ] 86.90% 74 eta 0s       \r",
      "[=========================================================>       ] 88.10% 75 eta 0s       \r",
      "[==========================================================>      ] 89.29% 76 eta 0s       \r",
      "[==========================================================>      ] 90.48% 77 eta 0s       \r",
      "[===========================================================>     ] 91.67% 78 eta 0s       \r",
      "[============================================================>    ] 92.86% 79 eta 0s       \r",
      "[=============================================================>   ] 94.05% 80 eta 0s       \r",
      "[=============================================================>   ] 95.24% 81 eta 0s       \r",
      "[==============================================================>  ] 96.43% 82 eta 0s       \r",
      "[===============================================================> ] 97.62% 83 eta 0s       \r",
      "[================================================================>] 98.81% 84 eta 0s       \r",
      "[=================================================================] 100.00% 85 0s 1ms\r\n",
      "Clustering step 2\r\n",
      "[>                                                                ] 0.00% 1 eta -\r",
      "[>                                                                ] 1.22% 2 eta 0s       \r",
      "[=>                                                               ] 2.44% 3 eta 0s       \r",
      "[==>                                                              ] 3.66% 4 eta 0s       \r",
      "[===>                                                             ] 4.88% 5 eta 0s       \r",
      "[===>                                                             ] 6.10% 6 eta 0s       \r",
      "[====>                                                            ] 7.32% 7 eta 0s       \r",
      "[=====>                                                           ] 8.54% 8 eta 0s       \r",
      "[======>                                                          ] 9.76% 9 eta 0s       \r",
      "[=======>                                                         ] 10.98% 10 eta 0s       \r",
      "[=======>                                                         ] 12.20% 11 eta 0s       \r",
      "[========>                                                        ] 13.41% 12 eta 0s       \r",
      "[=========>                                                       ] 14.63% 13 eta 0s       \r",
      "[==========>                                                      ] 15.85% 14 eta 0s       \r",
      "[===========>                                                     ] 17.07% 15 eta 0s       \r",
      "[===========>                                                     ] 18.29% 16 eta 0s       \r",
      "[============>                                                    ] 19.51% 17 eta 0s       \r",
      "[=============>                                                   ] 20.73% 18 eta 0s       \r",
      "[==============>                                                  ] 21.95% 19 eta 0s       \r",
      "[===============>                                                 ] 23.17% 20 eta 0s       \r",
      "[===============>                                                 ] 24.39% 21 eta 0s       \r",
      "[================>                                                ] 25.61% 22 eta 0s       \r",
      "[=================>                                               ] 26.83% 23 eta 0s       \r",
      "[==================>                                              ] 28.05% 24 eta 0s       \r",
      "[===================>                                             ] 29.27% 25 eta 0s       \r",
      "[===================>                                             ] 30.49% 26 eta 0s       \r",
      "[====================>                                            ] 31.71% 27 eta 0s       \r",
      "[=====================>                                           ] 32.93% 28 eta 0s       \r",
      "[======================>                                          ] 34.15% 29 eta 0s       \r",
      "[======================>                                          ] 35.37% 30 eta 0s       \r",
      "[=======================>                                         ] 36.59% 31 eta 0s       \r",
      "[========================>                                        ] 37.80% 32 eta 0s       \r",
      "[=========================>                                       ] 39.02% 33 eta 0s       \r",
      "[==========================>                                      ] 40.24% 34 eta 0s       \r",
      "[==========================>                                      ] 41.46% 35 eta 0s       \r",
      "[===========================>                                     ] 42.68% 36 eta 0s       \r",
      "[============================>                                    ] 43.90% 37 eta 0s       \r",
      "[=============================>                                   ] 45.12% 38 eta 0s       \r",
      "[==============================>                                  ] 46.34% 39 eta 0s       \r",
      "[==============================>                                  ] 47.56% 40 eta 0s       \r",
      "[===============================>                                 ] 48.78% 41 eta 0s       \r",
      "[================================>                                ] 50.00% 42 eta 0s       \r",
      "[=================================>                               ] 51.22% 43 eta 0s       \r",
      "[==================================>                              ] 52.44% 44 eta 0s       \r",
      "[==================================>                              ] 53.66% 45 eta 0s       \r",
      "[===================================>                             ] 54.88% 46 eta 0s       \r",
      "[====================================>                            ] 56.10% 47 eta 0s       \r",
      "[=====================================>                           ] 57.32% 48 eta 0s       \r",
      "[======================================>                          ] 58.54% 49 eta 0s       \r",
      "[======================================>                          ] 59.76% 50 eta 0s       \r",
      "[=======================================>                         ] 60.98% 51 eta 0s       \r",
      "[========================================>                        ] 62.20% 52 eta 0s       \r",
      "[=========================================>                       ] 63.41% 53 eta 0s       \r",
      "[==========================================>                      ] 64.63% 54 eta 0s       \r",
      "[==========================================>                      ] 65.85% 55 eta 0s       \r",
      "[===========================================>                     ] 67.07% 56 eta 0s       \r",
      "[============================================>                    ] 68.29% 57 eta 0s       \r",
      "[=============================================>                   ] 69.51% 58 eta 0s       \r",
      "[=============================================>                   ] 70.73% 59 eta 0s       \r",
      "[==============================================>                  ] 71.95% 60 eta 0s       \r",
      "[===============================================>                 ] 73.17% 61 eta 0s       \r",
      "[================================================>                ] 74.39% 62 eta 0s       \r",
      "[=================================================>               ] 75.61% 63 eta 0s       \r",
      "[=================================================>               ] 76.83% 64 eta 0s       \r",
      "[==================================================>              ] 78.05% 65 eta 0s       \r",
      "[===================================================>             ] 79.27% 66 eta 0s       \r",
      "[====================================================>            ] 80.49% 67 eta 0s       \r",
      "[=====================================================>           ] 81.71% 68 eta 0s       \r",
      "[=====================================================>           ] 82.93% 69 eta 0s       \r",
      "[======================================================>          ] 84.15% 70 eta 0s       \r",
      "[=======================================================>         ] 85.37% 71 eta 0s       \r",
      "[========================================================>        ] 86.59% 72 eta 0s       \r",
      "[=========================================================>       ] 87.80% 73 eta 0s       \r",
      "[=========================================================>       ] 89.02% 74 eta 0s       \r",
      "[==========================================================>      ] 90.24% 75 eta 0s       \r",
      "[===========================================================>     ] 91.46% 76 eta 0s       \r",
      "[============================================================>    ] 92.68% 77 eta 0s       \r",
      "[=============================================================>   ] 93.90% 78 eta 0s       \r",
      "[=============================================================>   ] 95.12% 79 eta 0s       \r",
      "[==============================================================>  ] 96.34% 80 eta 0s       \r",
      "[===============================================================> ] 97.56% 81 eta 0s       \r",
      "[================================================================>] 98.78% 82 eta 0s       \r",
      "[=================================================================] 100.00% 83 0s 2ms\r\n",
      "Write merged clustering\r\n",
      "[>                                                                ] 0.00% 1 eta -\r",
      "[>                                                                ] 1.19% 9 eta 0s       \r",
      "[=>                                                               ] 2.23% 16 eta 0s       \r",
      "[=>                                                               ] 2.37% 17 eta 0s       \r",
      "[==>                                                              ] 3.41% 24 eta 0s       \r",
      "[>                                                                ] 1.34% 10 eta 0s       \r",
      "[==>                                                              ] 4.60% 32 eta 0s       \r",
      "[===>                                                             ] 4.75% 33 eta 0s       \r",
      "[===>                                                             ] 5.79% 40 eta 0s       \r",
      "[>                                                                ] 1.04% 8 eta 0s       \r",
      "[====>                                                            ] 6.82% 47 eta 0s       \r",
      "[==>                                                              ] 4.45% 31 eta 0s       \r",
      "[====>                                                            ] 6.97% 48 eta 0s       \r",
      "[===>                                                             ] 5.64% 39 eta 0s       \r",
      "[=====>                                                           ] 8.01% 55 eta 0s       \r",
      "[=====>                                                           ] 8.16% 56 eta 0s       \r",
      "[=====>                                                           ] 9.20% 63 eta 0s       \r",
      "[======>                                                          ] 9.35% 64 eta 0s       \r",
      "[======>                                                          ] 9.50% 65 eta 0s       \r",
      "[======>                                                          ] 10.39% 71 eta 0s       \r",
      "[======>                                                          ] 10.53% 72 eta 0s       \r",
      "[=====>                                                           ] 9.05% 62 eta 0s       \r",
      "[======>                                                          ] 10.68% 73 eta 0s       \r",
      "[===>                                                             ] 5.93% 41 eta 0s       \r",
      "[=======>                                                         ] 11.57% 79 eta 0s       \r",
      "[=======>                                                         ] 11.87% 81 eta 0s       \r",
      "[========>                                                        ] 12.61% 86 eta 0s       \r",
      "[===>                                                             ] 6.08% 42 eta 0s       \r",
      "[========>                                                        ] 12.76% 87 eta 0s       \r",
      "[=======>                                                         ] 11.72% 80 eta 0s       \r",
      "[========>                                                        ] 13.80% 94 eta 0s       \r",
      "[==>                                                              ] 3.26% 23 eta 0s       \r",
      "[========>                                                        ] 12.91% 88 eta 0s       \r",
      "[=========>                                                       ] 15.28% 104 eta 0s       \r",
      "[==========>                                                      ] 16.32% 111 eta 0s       \r",
      "[========>                                                        ] 13.65% 93 eta 0s       \r",
      "[=========>                                                       ] 15.13% 103 eta 0s       \r",
      "[===========>                                                     ] 17.36% 118 eta 0s       \r",
      "[===========>                                                     ] 17.51% 119 eta 0s       \r",
      "[=========>                                                       ] 14.84% 101 eta 0s       \r",
      "[=========>                                                       ] 14.99% 102 eta 0s       \r",
      "[============>                                                    ] 18.55% 126 eta 0s       \r",
      "[============>                                                    ] 18.69% 127 eta 0s       \r",
      "[============>                                                    ] 19.73% 134 eta 0s       \r",
      "[=============>                                                   ] 20.62% 140 eta 0s       \r",
      "[==============>                                                  ] 21.66% 147 eta 0s       \r",
      "[==============>                                                  ] 22.70% 154 eta 0s       \r",
      "[===============>                                                 ] 23.74% 161 eta 0s       \r",
      "[===========>                                                     ] 17.80% 121 eta 0s       \r",
      "[=============>                                                   ] 20.77% 141 eta 0s       \r",
      "[================>                                                ] 24.78% 168 eta 0s       \r",
      "[================>                                                ] 25.07% 170 eta 0s       \r",
      "[===========>                                                     ] 17.66% 120 eta 0s       \r",
      "[============>                                                    ] 19.58% 133 eta 0s       \r",
      "[================>                                                ] 26.11% 177 eta 0s       \r",
      "[================>                                                ] 24.93% 169 eta 0s       \r",
      "[=================>                                               ] 27.15% 184 eta 0s       \r",
      "[=============>                                                   ] 20.92% 142 eta 0s       \r",
      "[=================>                                               ] 27.30% 185 eta 0s       \r",
      "[=================>                                               ] 26.41% 179 eta 0s       \r",
      "[=================>                                               ] 26.26% 178 eta 0s       \r",
      "[==================>                                              ] 28.34% 192 eta 0s       \r",
      "[==================>                                              ] 28.49% 193 eta 0s       \r",
      "[===================>                                             ] 29.53% 200 eta 0s       \r",
      "[===========>                                                     ] 18.40% 125 eta 0s       \r",
      "[===================>                                             ] 29.67% 201 eta 0s       \r",
      "[===================>                                             ] 29.82% 202 eta 0s       \r",
      "[====================>                                            ] 30.86% 209 eta 0s       \r",
      "[===================>                                             ] 30.71% 208 eta 0s       \r",
      "[====================>                                            ] 31.75% 215 eta 0s       \r",
      "[==================>                                              ] 28.19% 191 eta 0s       \r",
      "[===================>                                             ] 30.56% 207 eta 0s       \r",
      "[====================>                                            ] 31.90% 216 eta 0s       \r",
      "[=====================>                                           ] 32.79% 222 eta 0s       \r",
      "[=====================>                                           ] 32.94% 223 eta 0s       \r",
      "[=====================>                                           ] 33.83% 229 eta 0s       \r",
      "[====================>                                            ] 31.01% 210 eta 0s       \r",
      "[======================>                                          ] 33.98% 230 eta 0s       \r",
      "[======================>                                          ] 35.01% 237 eta 0s       \r",
      "[======================>                                          ] 34.27% 232 eta 0s       \r",
      "[=======================>                                         ] 36.05% 244 eta 0s       \r",
      "[======================>                                          ] 34.12% 231 eta 0s       \r",
      "[=======================>                                         ] 36.35% 246 eta 0s       \r",
      "[=======================>                                         ] 36.20% 245 eta 0s       \r",
      "[========================>                                        ] 37.39% 253 eta 0s       \r",
      "[======================>                                          ] 34.87% 236 eta 0s       \r",
      "[==================>                                              ] 28.64% 194 eta 0s       \r",
      "[========================>                                        ] 37.09% 251 eta 0s       \r",
      "[=========================>                                       ] 38.87% 263 eta 0s       \r",
      "[========================>                                        ] 38.43% 260 eta 0s       \r",
      "[=========================>                                       ] 39.91% 270 eta 0s       \r",
      "[=========================>                                       ] 38.58% 261 eta 0s       \r",
      "[==========================>                                      ] 40.95% 277 eta 0s       \r",
      "[========================>                                        ] 37.54% 254 eta 0s       \r",
      "[===========================>                                     ] 41.99% 284 eta 0s       \r",
      "[===========================>                                     ] 43.03% 291 eta 0s       \r",
      "[============================>                                    ] 44.07% 298 eta 0s       \r",
      "[=============================>                                   ] 45.10% 305 eta 0s       \r",
      "[=============================>                                   ] 45.25% 306 eta 0s       \r",
      "[==========================>                                      ] 40.06% 271 eta 0s       \r",
      "[==============================>                                  ] 46.44% 314 eta 0s       \r",
      "[==========================>                                      ] 41.10% 278 eta 0s       \r",
      "[=============================>                                   ] 46.14% 312 eta 0s       \r",
      "[==============================>                                  ] 47.48% 321 eta 0s       \r",
      "[==============================>                                  ] 47.63% 322 eta 0s       \r",
      "[=========================>                                       ] 38.72% 262 eta 0s       \r",
      "[===============================>                                 ] 47.77% 323 eta 0s       \r",
      "[===============================>                                 ] 48.81% 330 eta 0s       \r",
      "[===============================>                                 ] 48.96% 331 eta 0s       \r",
      "[================================>                                ] 49.85% 337 eta 0s       \r",
      "[===============================>                                 ] 47.92% 324 eta 0s       \r",
      "[=================================>                               ] 50.89% 344 eta 0s       \r",
      "[==============================>                                  ] 46.29% 313 eta 0s       \r",
      "[=================================>                               ] 51.93% 351 eta 0s       \r",
      "[================================>                                ] 50.30% 340 eta 0s       \r",
      "[=================================>                               ] 52.08% 352 eta 0s       \r",
      "[==================================>                              ] 52.37% 354 eta 0s       \r",
      "[==================================>                              ] 53.41% 361 eta 0s       \r",
      "[===================================>                             ] 54.45% 368 eta 0s       \r",
      "[====================================>                            ] 55.49% 375 eta 0s       \r",
      "[================================>                                ] 50.45% 341 eta 0s       \r",
      "[================================>                                ] 50.15% 339 eta 0s       \r",
      "[=====================================>                           ] 56.97% 385 eta 0s       \r",
      "[====================================>                            ] 56.53% 382 eta 0s       \r",
      "[=====================================>                           ] 58.01% 392 eta 0s       \r",
      "[====================================>                            ] 56.68% 383 eta 0s       \r",
      "[=================================>                               ] 52.23% 353 eta 0s       \r",
      "[=====================================>                           ] 58.31% 394 eta 0s       \r",
      "[================================>                                ] 50.00% 338 eta 0s       \r",
      "[=====================================>                           ] 58.16% 393 eta 0s       \r",
      "[====================================>                            ] 56.82% 384 eta 0s       \r",
      "[======================================>                          ] 59.20% 400 eta 0s       \r",
      "[=======================================>                         ] 60.39% 408 eta 0s       \r",
      "[=======================================>                         ] 61.42% 415 eta 0s       \r",
      "[=======================================>                         ] 60.53% 409 eta 0s       \r",
      "[========================================>                        ] 61.72% 417 eta 0s       \r",
      "[=======================================>                         ] 60.24% 407 eta 0s       \r",
      "[========================================>                        ] 61.57% 416 eta 0s       \r",
      "[========================================>                        ] 62.91% 425 eta 0s       \r",
      "[========================================>                        ] 62.76% 424 eta 0s       \r",
      "[========================================>                        ] 63.06% 426 eta 0s       \r",
      "[========================================>                        ] 61.87% 418 eta 0s       \r",
      "[=========================================>                       ] 63.95% 432 eta 0s       \r",
      "[==========================================>                      ] 64.99% 439 eta 0s       \r",
      "[==========================================>                      ] 66.02% 446 eta 0s       \r",
      "[===========================================>                     ] 67.06% 453 eta 0s       \r",
      "[===========================================>                     ] 67.21% 454 eta 0s       \r",
      "[===========================================>                     ] 66.17% 447 eta 0s       \r",
      "[============================================>                    ] 68.25% 461 eta 0s       \r",
      "[=============================================>                   ] 69.29% 468 eta 0s       \r",
      "[=============================================>                   ] 69.44% 469 eta 0s       \r",
      "[=============================================>                   ] 70.47% 476 eta 0s       \r",
      "[=============================================>                   ] 70.33% 475 eta 0s       \r",
      "[==============================================>                  ] 71.51% 483 eta 0s       \r",
      "[==============================================>                  ] 71.66% 484 eta 0s       \r",
      "[===============================================>                 ] 72.85% 492 eta 0s       \r",
      "[================================================>                ] 73.89% 499 eta 0s       \r",
      "[============================================>                    ] 68.40% 462 eta 0s       \r",
      "[===============================================>                 ] 72.70% 491 eta 0s       \r",
      "[================================================>                ] 74.33% 502 eta 0s       \r",
      "[================================================>                ] 74.48% 503 eta 0s       \r",
      "[=================================================>               ] 75.52% 510 eta 0s       \r",
      "[=================================================>               ] 76.56% 517 eta 0s       \r",
      "[==================================================>              ] 77.60% 524 eta 0s       \r",
      "[===================================================>             ] 78.64% 531 eta 0s       \r",
      "[===================================================>             ] 79.67% 538 eta 0s       \r",
      "[====================================================>            ] 80.71% 545 eta 0s       \r",
      "[=====================================================>           ] 81.75% 552 eta 0s       \r",
      "[================================================>                ] 75.37% 509 eta 0s       \r",
      "[=====================================================>           ] 82.34% 556 eta 0s       \r",
      "[======================================================>          ] 83.38% 563 eta 0s       \r",
      "[======================================================>          ] 84.42% 570 eta 0s       \r",
      "[=======================================================>         ] 85.46% 577 eta 0s       \r",
      "[========================================================>        ] 86.50% 584 eta 0s       \r",
      "[========================================================>        ] 87.54% 591 eta 0s       \r",
      "[======================================================>          ] 84.57% 571 eta 0s       \r",
      "[========================================================>        ] 87.69% 592 eta 0s       \r",
      "[=========================================================>       ] 88.72% 599 eta 0s       \r",
      "[==========================================================>      ] 89.76% 606 eta 0s       \r",
      "[===============================================>                 ] 72.55% 490 eta 0s       \r",
      "[===========================================================>     ] 90.95% 614 eta 0s       \r",
      "[===========================================================>     ] 90.80% 613 eta 0s       \r",
      "[===========================================================>     ] 92.14% 622 eta 0s       \r",
      "[===========================================================>     ] 91.99% 621 eta 0s       \r",
      "[============================================================>    ] 93.18% 629 eta 0s       \r",
      "[============================================================>    ] 93.32% 630 eta 0s       \r",
      "[=============================================================>   ] 94.36% 637 eta 0s       \r",
      "[==============================================================>  ] 95.40% 644 eta 0s       \r",
      "[==============================================================>  ] 96.44% 651 eta 0s       \r",
      "[===============================================================> ] 97.48% 658 eta 0s       \r",
      "[================================================================>] 98.52% 665 eta 0s       \r",
      "[================================================================>] 99.55% 672 eta 0s       \r",
      "[=================================================================] 100.00% 675 0s 5ms\r\n",
      "Time for merging to clu_redundancy: 0h 0m 0s 0ms\r\n",
      "Time for processing: 0h 0m 0s 9ms\r\n",
      "createsubdb tmp/245840764643801654/clu_redundancy pdbDB tmp/245840764643801654/input_step_redundancy -v 3 --subdb-mode 1 \r\n",
      "\r\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\r\n",
      "Time for processing: 0h 0m 0s 4ms\r\n",
      "prefilter tmp/245840764643801654/input_step_redundancy tmp/245840764643801654/input_step_redundancy tmp/245840764643801654/pref_step0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 1 -k 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 1 --comp-bias-corr 0 --comp-bias-corr-scale 1 --diag-score 0 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 0 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 8 --compressed 0 -v 3 \r\n",
      "\r\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Query database size: 83 type: Aminoacid\n",
      "Estimated memory consumption: 977M\n",
      "Target database size: 83 type: Aminoacid\n",
      "Index table k-mer threshold: 154 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 100.00% 83 0s 7ms      \n",
      "Index table: Masked residues: 112\n",
      "Index table: fill\n",
      "[=================================================================] 100.00% 83 0s 2ms      \n",
      "Index statistics\n",
      "Entries:          7728\n",
      "DB size:          488 MB\n",
      "Avg k-mer size:   0.000121\n",
      "Top 10 k-mers\n",
      "    AYPWGT\t63\n",
      "    WAGGAT\t40\n",
      "    WVLHRR\t29\n",
      "    RKFGWK\t18\n",
      "    KFVGKM\t17\n",
      "    QAEHIR\t16\n",
      "    WAGGAS\t16\n",
      "    MIDGWV\t10\n",
      "    SGVRHM\t9\n",
      "    IYGVGN\t9\n",
      "Time for index table init: 0h 0m 0s 301ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 154\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 83\n",
      "Target db start 1 to 83\n",
      "[=================================================================] 100.00% 83 0s 7ms      \n",
      "\n",
      "1.435504 k-mers per position\n",
      "384 DB matches per sequence\n",
      "0 overflows\n",
      "0 queries produce too many hits (truncated result)\n",
      "16 sequences passed prefiltering per query sequence\n",
      "20 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step0: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 620ms\n",
      "align tmp/245840764643801654/input_step_redundancy tmp/245840764643801654/input_step_redundancy tmp/245840764643801654/pref_step0 tmp/245840764643801654/aln_step0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.95 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 0 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 83 type: Aminoacid\n",
      "Target database size: 83 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 83 0s 33ms     \n",
      "Time for merging to aln_step0: 0h 0m 0s 0ms\n",
      "1329 alignments calculated\n",
      "91 sequence pairs passed the thresholds (0.068473 of overall calculated)\n",
      "1.096385 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 41ms\n",
      "clust tmp/245840764643801654/input_step_redundancy tmp/245840764643801654/aln_step0 tmp/245840764643801654/clu_step0 --cluster-mode 2 --max-iterations 1000 --similarity-type 2 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Clustering mode: Greedy\n",
      "Total time: 0h 0m 0s 0ms\n",
      "\n",
      "Size of the sequence database: 83\n",
      "Size of the alignment database: 83\n",
      "Number of clusters: 80\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to clu_step0: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "mergeclusters pdbDB clusterDB tmp/245840764643801654/clu_redundancy tmp/245840764643801654/clu_step0 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 100.00% 83 0s 1ms      \n",
      "Clustering step 2\n",
      "[=================================================================] 100.00% 80 0s 2ms      \n",
      "Write merged clustering\n",
      "[=================================================================] 100.00% 675 0s 5ms      \n",
      "Time for merging to clusterDB: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 10ms\n"
     ]
    }
   ],
   "source": [
    "!mmseqs cluster pdbDB clusterDB tmp --min-seq-id 0.95 --cluster-mode 2 --cov-mode 1 -c 0.8"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "de98df7e-723d-45bc-8fc2-5e72e48d7e5f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "createtsv pdbDB pdbDB clusterDB clusters.tsv \r\n",
      "\r\n",
      "MMseqs Version:                 \t14.7e284\r\n",
      "First sequence as representative\tfalse\r\n",
      "Target column                   \t1\r\n",
      "Add full header                 \tfalse\r\n",
      "Sequence source                 \t0\r\n",
      "Database output                 \tfalse\r\n",
      "Threads                         \t8\r\n",
      "Compressed                      \t0\r\n",
      "Verbosity                       \t3\r\n",
      "\r\n",
      "Time for merging to clusters.tsv: 0h 0m 0s 0ms\r\n",
      "Time for processing: 0h 0m 0s 5ms\r\n"
     ]
    }
   ],
   "source": [
    "!mmseqs createtsv pdbDB pdbDB clusterDB clusters.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "cca5508f-a859-4a55-a631-c75eee3c1233",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cleaning up files and directories we do not need\n",
    "!rm pdbDB*\n",
    "!rm -r tmp\n",
    "!rm clusterDB*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b9cc183-1429-49b7-8d50-2a7c1e8b8cd4",
   "metadata": {},
   "source": [
    "## Interpreting Our Clustering\n",
    "Next, we will generate **FASTA files** for each cluster and extract additional attributes.  \n",
    "\n",
    "- Below, we define helper **functions** to perform these tasks efficiently."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "9f61e469-dc1e-481c-8823-c17271afe857",
   "metadata": {},
   "outputs": [],
   "source": [
    "# run the code block to define the function\n",
    "def map_unique(list1, list2):\n",
    "    \"\"\"\n",
    "    Creates a mapping from each unique element in list2 to a set of corresponding elements from list1.\n",
    "\n",
    "    list1 : A list of values to be grouped.\n",
    "    list2 : A list of keys which will be used for grouping.\n",
    "    \"\"\"\n",
    "    zip_values = zip(list1, list2)\n",
    "    values_map = {}\n",
    "    \n",
    "    if len(list1) != len(list2):\n",
    "        return print('Both lists must be of equal lengths.')\n",
    "    \n",
    "    else:\n",
    "        for value1, value2 in zip_values:\n",
    "            if value2 not in values_map:\n",
    "                values_map[value2] = set()\n",
    "            values_map[value2].add(value1)\n",
    "        \n",
    "    return values_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "149c0a9e-7af1-4d9c-a5d9-44032cd0ea0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def cleaning_clusters(file_path):\n",
    "    \"\"\"\n",
    "    Creates a dictionary with unique cluster representatives as keys and grouped members as values.\n",
    "    \n",
    "    file_path : path of .tsv file containing information on cluster representatives and members.\n",
    "    \"\"\"\n",
    "    # fix indexing, extract whole thing\n",
    "    with open(file_path, 'r') as file:\n",
    "        label_list = []\n",
    "        representatives = []\n",
    "        members = []\n",
    "        for i, line in enumerate(file):\n",
    "            label_list.append(line)\n",
    "        for label in label_list:\n",
    "            representatives.append(label.split()[0])\n",
    "            members.append(label.split()[1])\n",
    "            \n",
    "        cluster_dict = map_unique(members, representatives)\n",
    "    return cluster_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "497862ce-ee80-4ae4-827b-95dd14d0fc80",
   "metadata": {},
   "outputs": [],
   "source": [
    "def writing_cluster_FASTA(seqres_dict, cluster_dict, output_directory):\n",
    "    \"\"\"\n",
    "    Created FASTA files out of the clusters made using MMSeqs2.\n",
    "\n",
    "    seqres_dict: dictionary made out of our seqres file utalizing seqres_entry_dict function\n",
    "    cluster_dict: Dictionary built off of the clusters.tsv having used the cleaning_clusters function\n",
    "    output_directory: directory where outputted files will be stored\n",
    "    \"\"\"\n",
    "    for key in cluster_dict:\n",
    "        with open(f'{output_directory}/{key}_cluster.txt', 'w') as file:\n",
    "            entries_list = list(cluster_dict[key])\n",
    "            for entry in entries_list:\n",
    "                lines_to_write = seqres_dict[entry]\n",
    "                if lines_to_write:  \n",
    "                    file.writelines(lines_to_write)\n",
    "                else:\n",
    "                    print(f'{entry} not in seqres.txt')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f25573bc-6979-4995-bbb8-f371b9b4190c",
   "metadata": {},
   "source": [
    "- **Task**: We create a dictionary out of our 'clusters.tsv' file. (read function documentation for more information)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "da2daceb-d873-431c-8a98-404ab620f706",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "An attribute of our cluster dictionary:\n",
      "5UJW_A : {'5UJW_E', '5UJW_D', '5UJW_F', '5UJW_A', '5UJW_C', '5UJW_B'}\n"
     ]
    }
   ],
   "source": [
    "cluster_dict = cleaning_clusters('clusters.tsv')\n",
    "\n",
    "cluster_key = list(cluster_dict.keys())[0]\n",
    "print(f'An attribute of our cluster dictionary:\\n{cluster_key} : {cluster_dict[cluster_key]}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a393b89c-13d9-48ae-b4b7-46a54c3561b9",
   "metadata": {},
   "source": [
    "- Additional information on our created clusters:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "40726681-b943-4b81-af3f-334108808558",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cluster member counts : 675\n",
      "cluster representatives count : 80\n"
     ]
    }
   ],
   "source": [
    "member_count = []\n",
    "for key in cluster_dict.keys():\n",
    "    member_count.append(len(cluster_dict[key]))\n",
    "\n",
    "print(f\"cluster member counts : {sum(member_count)}\")\n",
    "print(f'cluster representatives count : {len(cluster_dict.keys())}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b502c4c-4768-4d83-9544-f0e432407593",
   "metadata": {},
   "source": [
    "### Writing FASTA files out of our clusters"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22ed1c40-9c3a-4949-bf5d-bd9c30967860",
   "metadata": {},
   "source": [
    "We now run a function to generate **FASTA files** from our clusters and save them in the `cluster_files` directory."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "7687265d-5693-421f-bd79-7585ba9bcbf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "writing_cluster_FASTA(seqres_dict, cluster_dict, cluster_fasta_storage)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "feafee1d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 5 files in cluster_files:\n",
      "1AG1_O_cluster.txt\r\n",
      "1AW1_A_cluster.txt\r\n",
      "1B9B_A_cluster.txt\r\n",
      "1HG3_G_cluster.txt\r\n",
      "1I45_A_cluster.txt\r\n"
     ]
    }
   ],
   "source": [
    "print(f'First 5 files in {cluster_fasta_storage[:cluster_fasta_storage.index('/')]}:')\n",
    "!ls {cluster_fasta_storage}| head -n 5"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7cc2c84f-210b-4fa4-ac85-dff9694b61aa",
   "metadata": {},
   "source": [
    "### Writing our unique entity names file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1a02753-b42e-46fc-9dc0-0bd0b14e104d",
   "metadata": {},
   "source": [
    "In this step, we count every **unique entity** and identify the **most frequent entity** in our cluster FASTA files.  \n",
    "We append this information, along with the **total number of entities** across all clusters, into a `.txt` file.\n",
    "- Let's define functions needed for this task "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "c4681ddd-7148-42b4-a736-fcec5412fd05",
   "metadata": {},
   "outputs": [],
   "source": [
    "def obtaining_label_counts(file_path):\n",
    "    \"\"\"\n",
    "    Returns a dictionary, with the keys being the unique entity labels in a FASTA file \n",
    "    and the values being the counts of those keys in the file.\n",
    "\n",
    "    file_path : path to FASTA file\n",
    "    \"\"\"\n",
    "    label_list = get_attributes_seqres(file_path, 'ENTITY_LABEL')\n",
    "    unique_label_list = set(label_list)\n",
    "    counts = []\n",
    "    \n",
    "    for unique_label in unique_label_list:\n",
    "        count = 0\n",
    "        for label in label_list:\n",
    "            if label == unique_label:\n",
    "                count += 1\n",
    "        counts.append(count)\n",
    "    return dict(zip(unique_label_list, counts))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "68304044-3b18-400d-9a7a-06a38ded996a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def obtain_high_label(label_dict):\n",
    "    \"\"\"\n",
    "    Takes dictionary from obtaining_label_counts and returns a dictionary containing the key and value of \n",
    "    the most present entity label in the FASTA file\n",
    "\n",
    "    label_dict : dictionary from obtaining_label_counts()\n",
    "    \"\"\"\n",
    "    if len(label_dict.keys()) > 1:\n",
    "        label_dict = dict(sorted(label_dict.items(), key=lambda item: item[1], reverse=True))\n",
    "        return {list(label_dict.keys())[0] : label_dict[list(label_dict.keys())[0]]}\n",
    "    else:\n",
    "        return {}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "14f1b39e-bf46-47f1-b6c9-2a01b32af83f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def writing_labels_file(clusters_directory_path, output_directory):\n",
    "    \"\"\"\n",
    "    Creates a text file where each line contains the name of a cluster file, the most frequent \n",
    "    entity name for that cluster (along with their counts), and the unique entity names within \n",
    "    that cluster (also with its count).\n",
    "\n",
    "    clusters_directory_path : path to directory containing the cluster FASTA files\n",
    "\n",
    "    output_directory : directory where the newly created txt file will be stored\n",
    "    \"\"\"\n",
    "    lines_to_write = []\n",
    "    structure_count = 0\n",
    "\n",
    "    #This chunk removes a file we dont need.\n",
    "    ds_store_path = os.path.join(clusters_directory_path, '.DS_Store')\n",
    "    if os.path.exists(ds_store_path):\n",
    "        os.remove(ds_store_path)\n",
    "        \n",
    "    for file in os.listdir(clusters_directory_path):\n",
    "            \n",
    "        full_path = os.path.join(clusters_directory_path, file)\n",
    "\n",
    "        #This line counts how main chains are in the current file its reading, \n",
    "        # and adds up that total to its count. It repeats that for every file in the cluster\n",
    "        structure_count += len(get_attributes_seqres(full_path, 'chid'))\n",
    "        test_dict = obtaining_label_counts(full_path)\n",
    "        high_label = obtain_high_label(test_dict)\n",
    "            \n",
    "        if high_label:\n",
    "            line = [f'{file}',f'{high_label}',f'{test_dict}']\n",
    "        else:\n",
    "            line = [f'{file}',f'{test_dict}']\n",
    "            \n",
    "        lines_to_write.append(line)\n",
    "    \n",
    "    with open(f'{output_directory}/unique_entity_names.txt', 'w') as curr_file:\n",
    "        curr_file.writelines(f'TOTAL CHAINS : {structure_count}\\n\\n')\n",
    "        for line in lines_to_write:\n",
    "            curr_file.writelines(line)\n",
    "            curr_file.writelines('\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59f3720f-7fc9-42b4-ae2d-422c3e115af0",
   "metadata": {},
   "source": [
    "- **Task:** creating our unique entity names file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "52e2602d-7dea-485a-9e8c-c5b0aeecd27f",
   "metadata": {},
   "outputs": [],
   "source": [
    "writing_labels_file(cluster_fasta_storage, '.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "f8cb40cf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 5 lines of our unique_entity_names.txt\n",
      "TOTAL CHAINS : 675\r\n",
      "\r\n",
      "1HG3_G_cluster.txt{'TRIOSEPHOSPHATE ISOMERASE': 8}\r\n",
      "4Y90_A_cluster.txt{'TRIOSEPHOSPHATE ISOMERASE': 4}\r\n",
      "1I45_A_cluster.txt{'TRIOSEPHOSPHATE ISOMERASE': 16}\r\n"
     ]
    }
   ],
   "source": [
    "print('First 5 lines of our unique_entity_names.txt')\n",
    "!head -n 5 unique_entity_names.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51c8e881",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (task_test)",
   "language": "python",
   "name": "task_test"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
